src/connectivity.jl

```diff
@@ -129,9 +135,69 @@ julia> is_connected(g)
 true
 ```
 """
-function is_connected(g::AbstractGraph)
+function is_connected(g::AbstractGraph{T}) where {T}
     mult = is_directed(g) ? 2 : 1
-    return mult * ne(g) + 1 >= nv(g) && length(connected_components(g)) == 1
+    if mult * ne(g) + 1 >= nv(g)
+        label = zeros(T, nv(g))
+        connected_components!(label, g)
+        return allequal(label)
+    else
+        return false
+    end
+end
+
+"""
+    count_connected_components( g, [label, search_queue]; reset_label::Bool=false)
+
+Return the number of connected components in `g`.
+
+Equivalent to `length(connected_components(g))` but uses fewer allocations by not
+materializing the component vectors explicitly. Additionally, mutated work-arrays `label`
+and `search_queue` can be provided to reduce allocations further (see
+[`connected_components!`](@ref)).
+
+## Keyword arguments
+- `reset_label :: Bool` (default, `false`): if `true`, `label` is reset to zero before
+  returning.
+
+## Example
+```
+julia> using Graphs
+
+julia> g = Graph(Edge.([1=>2, 2=>3, 3=>1, 4=>5, 5=>6, 6=>4, 7=>8]));
+
+length> connected_components(g)
+3-element Vector{Vector{Int64}}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8]
+
+julia> count_connected_components(g)
+3
+```
+"""
+function count_connected_components(
+    g::AbstractGraph{T},
+    label::AbstractVector{T}=zeros(T, nv(g)),
+    search_queue::Vector{T}=Vector{T}();
+    reset_label::Bool=false,
+) where {T}
+    connected_components!(label, g, search_queue)
+    c = count_unique(label)
+    reset_label && fill!(label, zero(eltype(label)))
+    return c
+end
+
+function count_unique(label::Vector{T}) where {T}
+    seen = Set{T}()
+    c = 0
+    for l in label
+        if l ∉ seen
+            push!(seen, l)
+            c += 1
+        end
+    end
+    return c
```

Comment:
```suggestion
    return length(Set(label))
```

Reply:
That's less performant than the explicit looped version though:

```jl
julia> label_small = rand(1:3, 20)
julia> @b count_unique($label_small)
150.851 ns (4 allocs: 320 bytes) # loop
174.412 ns (4 allocs: 464 bytes) # length(Set(label))

julia> label_big = rand(1:50, 5000)
julia> @b count_unique($label_big)
23.385 μs (11 allocs: 3.312 KiB) # loop
32.719 μs (6 allocs: 72.172 KiB) # length(Set(label))

julia> label_huge = rand(1:5000, 500000)
julia> @b count_unique($label_huge)
3.499 ms (25 allocs: 192.625 KiB) # loop
4.876 ms (6 allocs: 9.000 MiB, 2.51% gc time)  # length(Set(label))
```

Reply:
It's indeed not very great that the `length(Set(label))` version is slower though. The reasons seems to be that `Set(itr)` is assuming that most elements in `itr` will be unique and goes ahead an `sizehint!`s the to-be-filled-in `Set` to be the full length of `itr` - but that seems very unlikely to ever be the case in this scenario: there will usually be far fewer connected components than vertices.

Reply:
Actually, it is not really an "issue" in Base, per se: rather, it seems `Set` is optimized with the assumption that most things that are `push!`ed into it are new, unique things. But when that assumption doesn't apply, it is faster to check `∉` before trying to `push!`. Here, I would say it's very safe to assume that `label` will usually contain far fewer unique things than its length, so we might as well exploit that.

Reply:
That's interesting. I did not know that. Btw. if try to be really efficient here - would using `BitSet` instead of `Set` be even more efficient?